### PR TITLE
Use MaxRect algorithm to pack textures; add dropdown to atlas preview window to see changes

### PR DESCRIPTION
**Pull request by matyasatfp**
Same as my previous PR, just in one commit

### CODE REVIEW

**Comment by tconkling** [comment]
Cool. I will take a look at this in the next few days, and run it against my own project. 

@roguenet and @aduros, do you guys also want to take a look?

**Comment by tconkling** [comment]
Move to https://github.com/threerings/flump/pull/135

**Comment by tconkling** [comment]
(Whoops, I closed the wrong pull request)

**Comment by roguenet** [comment]
Yeah, I'll take a look and see about running it against the MM project
files - I'm interested to see if this algorithm takes care of the
ridiculous slowness on our level files we saw with the old one.

On Tue, Jan 13, 2015 at 5:44 PM, Tim Conkling <notifications@github.com>
wrote:

> (Whoops, I closed the wrong pull request)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/threerings/flump/pull/135#issuecomment-69856075>.
>



-- 
Nathan Curtis

**Comment by aduros** [comment]
I don't think I'll have time to try it out this week, so feel free to merge.

File: exporter/src/main/as/flump/export/texturepacker/MaxRectMultiPacker.as [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+import flump.SwfTexture;
+import flump.Util;
+import flump.export.Atlas;
+import flump.export.AtlasImpl;
+
+// Packer that tries to use as few atlases as possible
+public class MaxRectMultiPacker extends MultiPackerBase {
+
+    // TODO the last 3 parameters are not needed for this class
```
Comment (matyasf) [inline exporter/src/main/as/flump/export/texturepacker/MaxRectMultiPacker.as]: removing them from here would needed a refactoring, I did not want to do this

File: exporter/src/main/as/flump/export/TexturePacker.as [hunk 2]
```diff
@@ -1,333 +0,0 @@
...
-            Preconditions.checkState(_unpacked.length == 0 || !hasEmptyAtlas,
-                "Texture won't fit in newly-created atlas?");
-        }
-
-        var totalTime :int = getTimer() - start;
-        log.info("Finished packing", "quality", quality, "scale", scale, "time", totalTime / 1000);
-    }
-
-    // Estimate the optimal size for the next atlas
-    protected function findOptimalSize () :Point {
```
Comment (matyasatfp) [inline exporter/src/main/as/flump/export/TexturePacker.as]: This function is moved to TexturePackerBase

**Comment by matyasatfp** [comment]
Are you planning to review/merge this? Or is Flump abandoned? I think I'll need to make other improvements. So now I either care about my fork only or face big headaches because of pull request merge conflicts.

**Comment by tconkling** [comment]
Sorry for the late response, matyasaftp. I've been head-down on my game and have been neglecting your pull request. Checking it out now!

**Comment by matyasf** [comment]
No problem, in the meanwhile ive made improvements in my own branch. The 2 main ones are:
- Much faster instantiation (parsing a flump .zip is around 2x faster and instantiating a parsed movie is around 9x faster)
- Assets are exported based on the maximum size you use them. Thus artists can scale them however they want and they wont look pixelated or take up unneccesary space.

This PR has a bug, you can see the fix in https://github.com/funkypandagame/flump/commit/407b8e4a434c05c3f46c81b51f77e662e61b66d8 MaxRectMultiPacker.as. I suggest if you merge then make another commit with these changes.

**Comment by tconkling** [comment]
I rebased your commits rather than merging them (which, in hindsight, was probably the wrong thing to do as Github does not know that these commits are part of mainline).

Thanks again!

**Comment by markknol** [comment]
@matyasf That also means that you cannot optimized assets. for example, we use a lot of 10x10px sprites that get stretched to fill backgrounds for example. Also for particle effects we use smaller assets which get scaled bigger to have better performance. It's very nice to be in control which stuff gets scaled and which not. Our artist can perfectly deal with this scaling behavior, when they get instructed to keep it in mind. I hope this gets implemented as a setting.

**Comment by tconkling** [comment]
@matyasatfp, can you submit your bug fixes as a pull request?

**Comment by tconkling** [comment]
@matyasatfp I'd definitely be interested in merging your movie instantiation optimizations. I share @markknol's trepidations about the auto-scaling behavior, which I think would only work as an optional setting.

**Comment by tconkling** [comment]
Also: @matyasatfp, I'm pulling your bugfix from 407b8e4a434c05c3f46c81b51f77e662e61b66d8, though the only way I can do this cleanly is with a cherry-pick commit, which means that our repos will be diverging again. If you want to submit more pull requests, it probably makes sense to rebase them off a new clean branch off threerings flump/master

**Pull request closed without merging**